lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+		$this->attribute_end   = $end;
+	}
+}
+
+
+class WP_HTML_Scanner {
+	/**
+	 * Input HTML document we're processing.
+	 *
+	 * @var string
+	 */
+	public $document;
+
+
+	/**
+	 * Byte offset in the input document we will start or continue parsing.
+	 *
+	 * @var int
+	 */
+	public $start_at = 0;
+
+
+	/**
+	 * Byte offset in the current tag being inspected starts, includes leading `<`.
+	 *
+	 * @var int|null
+	 */
+	public $tag_start = null;
+
+
+	/**
+	 * Start and end of matched tag name, or `null` if not yet found.
+	 *
+	 * @var WP_HTML_Scanner_Token|null
+	 */
+	public $tag_name = null;
+
+	/**
+	 * Lazily-built index of found attributes within an HTML tag match.
+	 *
+	 * Example:
+	 * <code>
+	 *     // supposing the parser is working through this content
+	 *     // and stops after recognizing the `id` attribute
+	 *     // <div id="test-4" class=outline title="data:text/plain;base64=asdk3nk1j3fo8">
+	 *     //                 ^ parsing will continue from this point
+	 *     $this->attributes = array(
+	 *         'id' => array( 6, 17, 'test-4' )
+	 *     );
+	 *
+	 *     // when picking up parsing again, or when asking to find the
+	 *     // `class` attribute we will continue and add to this array
+	 *     $this->attributes = array(
+	 *         'id' => array( 6, 17, 'test-4' ),
+	 *         'class' => array( 18, 32, 'outline' )
+	 *     );
+	 * </code>
+	 *
+	 * @var array<WP_HTML_Attribute_Token>
+	 */
+	public $attributes = array();
+
+
+	public function __construct( $input ) {
+		$this->document = $input;
+	}
+
+
+	public function scan( WP_Tag_Find_Descriptor $descriptor, $found_already = 0, $tag = null ) {
+		if ( $found_already === $descriptor->match_offset ) {
+			return $tag;
+		}
+
+		$tag = $this->find_next_tag();
+		if ( ! $tag ) {
+			return false;
+		}
+
+		// @TODO: Are we done matching?
+		while ( $attribute = $this->find_next_attribute() ) {
+			// HTML5 says duplicate values are ignored
+			if ( ! array_key_exists( $attribute->name->comparable, $this->attributes ) ) {
+				$this->attributes[ $attribute->name->comparable ] = $attribute;
+			}
+		}
+
+		$does_match = 'matches' === $descriptor->check( $tag->comparable, $this->attributes );
+		$found_this_time = $does_match ? 1 : 0;
+
+		return $this->scan( $descriptor, $found_already + $found_this_time, $tag );
+	}
+
+	/**
+	 * @return WP_HTML_Scanner_Token|false
+	 */
+	public function find_next_tag() {
+		// @TODO: Handle <DOCTYPE>
+		if ( 1 !== preg_match(
+			/*
+			 * Unfortunately we can't try to search for only the tag name we want because that might
+			 * lead us to skip over other tags and lose track of our place. So we need to search for
+			 * _every_ tag and then check after we find one if it's the one we are looking for.
+			 */
+	"~<!--(?>.*?-->)|<!\[CDATA\[(?>.*?>)|<\?(?>.*?)>|<(?P<TAG>[a-z][^\t\x{0A}\x{0C} /?>]*)~Smui",
+			$this->document,
+			$tag_match,
+			PREG_OFFSET_CAPTURE,
+			$this->start_at
+		) ) {
+			return false;
```

Comment:
We need to reset the state if there's no match:

```suggestion
			$this->start_at = null;
			$this->tag_name = null;
			$this->tag_start = null;
			$this->attributes = array();
			return false;
```

Otherwise the following code adds two attributes to the span:

```php
$p = new WP_HTML_Processor( '<span/>' );
$p->find( ['tag_name'=>'span'] );
$p->set_attribute( 'attr-1', 'a' );
$p->find( ['tag_name'=>'div'] );
$p->set_attribute( 'attr-2', 'a' );
echo $p;
```

Reply:
very good catch. I will update this